JavaScript development/Frameworks/React.md

```diff
@@ -0,0 +1,598 @@
+        })}
+      />
+    );
+  }
+}
+```
+
+## React + MobX
+*For general information about MobX, checkout the official documentation:* [`mobx.js.org`](https://mobx.js.org/).
+
+### Decorators
+Use decorators wherever you can if your transpiler support's it, because it makes code more readable and declarative.
+
+### Component Declaration
+Wrap components that handle app's state with *observer* from the `mobx-react` library.
+
+``` jsx
+import {observer} from 'mobx-react';
+// ...
+@observer
+export class ComponentName extends React.Component {
+  // ...
+}
+```
+
+### Local Component State
+If MobX is used for state management:
+- preffer using MobX *observable* instead of *this.state* for component's local state; don't use both
+- wrap local state in a single observable object named **componentState**
+
+``` jsx
+import {observable} from 'mobx';
+import {observer} from 'mobx-react';
+// ...
+@observer
+export class ComponentName extends React.Component {
+  // ...
+  @observable componentState = {
+    selectedTitle: 'React+MobX',
+  };
+
+  render() {
+    // ...
+    return (
+      <div>
+        // ...
+        <div>{this.componentState.selectedTitle}</div>
+      </div>
+    );
+  }
+}
+```
+
+### Computed Values
+If you need values derived from observable data, use *@computed*.
+
+``` jsx
+import {observable, computed} from 'mobx';
+import {observer} from 'mobx-react';
+// ...
+
+@observer
+export class ComponentName extends React.Component {
+  // ...
+  @observable componentState = {
+    firstName: 'John',
+    lastName: 'Doe',
+  };
+
+  @computed get fullName() {
+    return `${this.componentState.firstName} ${this.componentState.lastName}`;
+  }
+
+  render() {
+    return (
+      <div>{this.fullName}</div>
+    );
+  }
+}
+```
+
+### Actions
+Wrap methods and functions that modify state with *@action*. For binding action methods, use the aforementioned *@bind* decorator from the [`developit/decko`](https://github.com/developit/decko) lib.
+``` jsx
+  import {action, observable} from 'mobx';
+  import {observer} from 'mobx-react';
+// ...
+export class ComponentName extends React.Component {
+  // ...
+  @observable componentState = {
+    counter: 0,
+  };
+
+  public componentWillMount() {
+    // ...
+  }
+
+  // Note: lifecycle methods also have to be wrapped as "actions" if they modify state
+  @action public componentDidMount() {
+    // ...
+  }
+
+  @bind
+  @action
+  increaseCounter() {
+    this.componentState.counter++;
+  }
+
+  render() {
+    return (
+      <div>
+        <div>{this.componentState.counter}</div>
+        <button onClick={this.increaseCounter}>Increase by One</button>
+      </div>
+    );
+  }
+}
+```
+
+### Async Await
+In actions implemented as *async* functions, after each *await*, any code that modifies the state should be wrapped in a *runInAction* utility. This is because *async await* is just syntax sugar, and in reallity, a new function is created after each *await*.
```

Comment:
> This is because *async await* is just syntax sugar, and in reallity, a new function is created after each *await*.

This is only true for the transpiled code, but is also not relevant. The reason is that actions need to be synchronous.